lib/src/Uncertainty/Algorithm/Optimization/LOLAVoronoi.cxx

```diff
@@ -0,0 +1,417 @@
+        if (neighbourhoodScore > bestNeighbourhoodScore)
+        {
+          neighbourhoodI.clear();
+          for (UnsignedInteger j = 0; j < m; ++ j)
+            neighbourhoodI.add(candidateIndices[candidateCombinations_(n, j)]);
+          bestNeighbourhoodScore = neighbourhoodScore;
+        }
+
+      } // candidate neighbourhood loop
+
+      // now compute the score
+      const Point prx(lola_.x_[i]);
+      const Point pry(lola_.y_[i]);
+
+      Matrix p(m, d);
+      for (UnsignedInteger ti = 0; ti < m; ++ ti)
+      {
+        const Point prtx(lola_.x_[neighbourhoodI[ti]]);
+        for (UnsignedInteger j = 0; j < d; ++ j)
+          p(ti, j) = prtx[j] - prx[j];
+      }
+
+      Scalar eprAgg = 0.0;
+      for (UnsignedInteger k = 0; k < lola_.y_.getDimension(); ++ k)
+      {
+        Point f(m);
+        for (UnsignedInteger ti = 0; ti < m; ++ ti)
+        {
+          const Point prty(lola_.y_[neighbourhoodI[ti]]);
+          f[ti] = prty[k];
+        }
+
+        // g = \argmin ||pg-f|| cf 3.3.3 equation (3.8)
+        const Point g(LeastSquaresMethod::Build(methodName_, p).solve(f));
+
+        // local nonlinearity, 3.3.4 equation (3.9)
+        Scalar epr = 0.0;
+        for (UnsignedInteger ti = 0; ti < m; ++ ti)
+        {
+          const Point prtx(lola_.x_[neighbourhoodI[ti]]);
+          const Point prty(lola_.y_[neighbourhoodI[ti]]);
+          epr += std::abs(prty[k] - (pry[k] + g.dot(prtx - prx)));
+        }
+
+        // the non linearity score is the max across output components cf 3.6 equation (3.12)
+        if (aggregationMethod_ == "Maximum")
+          eprAgg = std::max(eprAgg, epr);
+        else if (aggregationMethod_ == "Average")
+          eprAgg += epr / lola_.y_.getDimension();
+        else
+          throw InvalidArgumentException(HERE) << "LOLAVoronoi-NonLinearityAggregationMethod must be 'Maximum' or 'Average'";
+      }
+      lola_.lolaScore_[i] = eprAgg;
+    } // i loop
+  }
+};
+
+void LOLAVoronoi::computeLOLAScore() const
+{
+  const UnsignedInteger previousSize = neighbourhood_.getSize();
+  neighbourhood_.resize(x_.getSize());
+
+  lolaScore_.resize(x_.getSize());
+  const LVLOLAScorePolicy policy(*this, previousSize);
+  TBBImplementation::ParallelFor(0, x_.getSize(), policy);
+}
+
+Sample LOLAVoronoi::generate(const UnsignedInteger size) const
+{
+  if (size > x_.getSize())
+    throw InvalidArgumentException(HERE) << "cannot ask more than " << x_.getSize() << " new samples (asked " << size << ")";
+
+  tree_ = KDTree(x_);
+
+  // hybrid score, see 3.4 equation (3.10)
+  Sample hybridScore(x_.getSize(), 1);
+  LOGINFO("LOLAVoronoi updating voronoi score");
+  computeVoronoiScore();
+  LOGINFO("LOLAVoronoi updating LOLA score");
+  computeLOLAScore();
+  const Scalar sumLS = std::accumulate(lolaScore_.begin(), lolaScore_.end(), 0.0);
+  for (UnsignedInteger i = 0; i < x_.getSize(); ++ i)
+    hybridScore(i, 0) = voronoiScore_[i] + lolaScore_[i] / sumLS;
```

Comment:
May this hybrid score be accessed to the user?

Reply:
yes, see the api